Extend give command "skill level" & shortening /talent all

natsurepo opened the pull request:
## Description

Attempting to add talents / skill level as a give command parameter, also shortening /talent all command
- [x] ~~Adding getCombatSkills for grabbing skill ids specifically for talent command, since getSkillsAndEnergySkill also grabs other unrelated skills~~ (removed since getSkillsAndEnergySkill is enough and there is a level cap for invalid talents)
- [x] Shorten /talent all code
- [x] Adding skillLevel as parameter so people can also specify talent levels for characters in /give
- [x] Resolving the issue for skillLevel parameter messing with constellations not giving extra skillLevel at the end (ex. at C6, talent E and Q doesn't increase by 3) (resolved by [1ecc3f4](https://github.com/Grasscutters/Grasscutter/commit/1ecc3f43905d4d29f409d86654aa113bc948b2af))
- [x] Cleanup 

A bit of note in the end, I would appreciate any suggestions and help for either resolving my current issue and making the implementation tidier.
<!--- Put the links of issues that may be fixed by this PR here (if any). -->
## Type of changes

<!--- Put an `x` in all the boxes that apply your changes. -->

- [ ] Bug fix
- [ ] New feature 
- [x] Enhancement
- [ ] Documentation

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] My pull request is unique and no other pull requests have been opened for these changes
- [x] I have read the [Contributing note](https://github.com/Grasscutters/Grasscutter/blob/stable/CONTRIBUTING.md) and [Code of conduct](https://github.com/Grasscutters/Grasscutter/blob/development/CODE_OF_CONDUCT.md)
- [x] I am responsible for any copyright issues with my code if it occurs in the future.

natsurepo committed:
Extend give command "talent"

natsurepo committed:
Merge remote-tracking branch 'origin/development' into development

Birdulon commented on src/main/java/emu/grasscutter/game/avatar/Avatar.java:
```diff
@@ -779,6 +779,15 @@ public boolean setSkillLevel(int skillId, int level) {
 
+    public void changeSkillLevel(int skillLevel) {
+        if (skillLevel < 1) skillLevel = 1;
+        if (skillLevel > 12) skillLevel = 12; // Set max to 12 at the moment instead of 15 (it will conflict with constellation's talent level bonus)
```
Check `Avatar::setSkillLevel`, constellation bonuses have also been capped in 4e75361 which also correctly caps other talents (e.g. alt sprints) so it would be safe (even if slightly incorrect) to just use `AvatarSkillDepotData::getSkillsAndEnergySkill` elsewhere.

natsurepo replied:
I added `Avatar::changeSkillLevel` specifically for my implementation since I only needed skillLevel as input (instead of also needing skillId), also using `Avatar::setSkillLevel` leads to `"emu.grasscutter.game.avatar.Avatar.getPlayer()" is null`.

Also for the `AvatarSkillDepotData::getSkillsAndEnergySkill` does it mean I should revert my changes for `AvatarSkillDepotData` and `TalentCommand`?

Birdulon replied:
>also using `Avatar::setSkillLevel` leads to `"emu.grasscutter.game.avatar.Avatar.getPlayer()" is null`.

Ah right, won't have a player during construction. Should be fine to add a conditional block to the packets in it.
```java
        // Packet
        val player = this.getPlayer();
        if (player != null) {
            player.sendPacket(new PacketAvatarSkillChangeNotify(this, skillId, oldLevel, level));
            player.sendPacket(new PacketAvatarSkillUpgradeRsp(this, skillId, oldLevel, level));
        }
```

> Also for the `AvatarSkillDepotData::getSkillsAndEnergySkill` does it mean I should revert my changes for `AvatarSkillDepotData` and `TalentCommand`?

Up to you, might be useful regardless and it's a pretty harmless addition.

> I added `Avatar::changeSkillLevel` specifically for my implementation since I only needed skillLevel as input 

This is a bit overspecialised though.

natsurepo replied:
```java
public boolean setSkillLevel(int skillId, int level) {
    if (level < 0 || level > 15) return false;
    var validLevels = GameData.getAvatarSkillLevels(skillId);
    if (validLevels != null && !validLevels.contains(level)) return false;
    int oldLevel = this.skillLevelMap.getOrDefault(skillId, 0);
    this.skillLevelMap.put(skillId, level);
    this.save();
    // Packet
    val player = this.getPlayer();
    if (player != null) {
        player.sendPacket(new PacketAvatarSkillChangeNotify(this, skillId, oldLevel, level));
        player.sendPacket(new PacketAvatarSkillUpgradeRsp(this, skillId, oldLevel, level));
    }
    return true;
}

public void changeSkillLevel(int skillLevel) {
    int finalSkillLevel = skillLevel;
    skillDepot.getCombatSkills().forEach(id -> this.setSkillLevel(id, skillLevel));
}
```

Made some changes with your suggestion and it works with `setSkillLevel` now. I still keep `changeSkillLevel` as an intermediate operation to fetch skillIds. Still no luck with bonus talent levels from the constellations though.

Birdulon commented on src/main/java/emu/grasscutter/data/excels/AvatarSkillDepotData.java:
```diff
@@ -77,4 +77,10 @@ public IntStream getSkillsAndEnergySkill() {
     }
+
+    public IntStream getCombatSkills() {
+        return IntStream.concat(IntStream.of(this.getSkills().get(0), this.getSkills().get(1)), IntStream.of(this.energySkill))
```
```suggestion
        return IntStream.of(this.getSkills().get(0), this.getSkills().get(1), this.energySkill)
```

natsurepo committed:
Update src/main/java/emu/grasscutter/data/excels/AvatarSkillDepotData.java

Shorten IntStream for getCombatSkills

Co-authored-by: Luke H-W <Birdulon@users.noreply.github.com>

natsurepo committed:
Fix setSkillLevel to work during avatar construction
Shortening getCombatSkills

natsurepo committed:
Merge remote-tracking branch 'origin/development' into development

natsurepo committed:
changeSkillLevel now acts as intermediate operation to fetch skillIds

Birdulon commented:
Added cdb66a74 after noticing the `sl(\d+)` regex was getting beaten out by the existing `l(\d+)` match for level

natsurepo committed:
Merge branch 'development' into development

natsurepo committed:
setSkillLevel changes to allow out of range levels to be normalized

Birdulon commented on src/main/java/emu/grasscutter/command/commands/GiveCommand.java:
```diff
@@ -212,30 +214,29 @@ public void execute(Player sender, Player targetPlayer, List<String> args) {
+        avatar.changeSkillLevel(skillLevel);
         avatar.forceConstellationLevel(constellation);
-        avatar.recalcStats();
+        avatar.recalcConstellations();
```
```suggestion
```
`Avatar::forceConstellationLevel` takes care of this. It also handles `recalcStats`, but it might not call that if it's fed an invalid constellation level so leaving `recalcStats` in is fine.

natsurepo replied:
Alright, thank you

natsurepo committed:
Update src/main/java/emu/grasscutter/command/commands/GiveCommand.java

Removing recalcStats since it's redundant

Co-authored-by: Luke H-W <Birdulon@users.noreply.github.com>

Birdulon commented on src/main/java/emu/grasscutter/game/avatar/Avatar.java:
```diff
@@ -762,11 +763,19 @@ public boolean setSkillLevel(int skillId, int level) {
         return true;
     }
 
+    public void changeSkillLevel(int skillLevel) {
```
Still not too keen on this method being added.

natsurepo replied:
Removed `changeSkillLevel`, added -1 as input for `skillId` in `GiveCommand` and `setSkillLevel` will fetch those ids within itself if `skillId` is -1

natsurepo replied:
I tried it in game, no conflicts with other usages too

Birdulon commented on src/main/java/emu/grasscutter/game/avatar/Avatar.java:
```diff
@@ -753,7 +753,8 @@ public boolean upgradeSkill(int skillId) {
     public boolean setSkillLevel(int skillId, int level) {
-        if (level < 0 || level > 15) return false;
+        if (level < 1) level = 1;
+        if (level >15) level = 15;
```
Shouldn't cap levels here, also setting to 0 is a feature that shouldn't be removed.

natsurepo replied:
```java
    private static void giveAllAvatars(Player player, GiveItemParameters param) {
        int promoteLevel = Avatar.getMinPromoteLevel(param.lvl);
        if (param.constellation < 0 || param.constellation > 6) param.constellation = 6; // constellation's default is -1 so if no parameters set for constellations it'll automatically be 6
        if (param.skillLevel > 15) param.skillLevel = 15;
        if (param.skillLevel < 0) param.skillLevel = 1; // Still allowing 0 to be used, while everything below 0 goes back to default
        for (AvatarData avatarData : GameData.getAvatarDataMap().values()) {
            int id = avatarData.getId();
            if (id < 10000002 || id >= 11000000) continue; // Exclude test avatars
            // Don't try to add each avatar to the current team
            player.addAvatar(makeAvatar(avatarData, param.lvl, promoteLevel, param.constellation, param.skillLevel), false);
        }
    }
```
Can I move it here instead or remove it completely?

Birdulon replied:
Just revert the two line change :)

natsurepo committed:
Major changes and cleanup:
- AvatarSkillDepotData: removed getCombatSkills since it's unused
- TalentCommand: shortened /talent all using getSkillsAndEnergySkill
- GiveCommand: changed changeSkillLevel to setSkillLevel
- Avatar: delete changeSkillLevel and moved the operation inside setSkillLevel,updated skillId to Integer to catch special cases from GiveCommand

natsurepo committed:
Merge remote-tracking branch 'origin/development' into development

natsurepo committed:
Small cleanup:
- GiveCommand: changed null to -1
- Avatar: changed null to -1 and revert skillId from Integer back to int

natsurepo committed:
Merge branch 'Grasscutters:development' into development

natsurepo committed:
Merge remote-tracking branch 'origin/development' into development

natsurepo committed:
Small cleanup:
Removed the special case from Avatar to be handled inside of GiveCommand
